### PR TITLE
[Issue #39] tab render optimization

### PR DESCRIPTION
**Pull request by cnojima**
# Pull Request

## Description
Current beta implementation mounts all views and controllers (and hooks and global refs, etc) at the same time.  State mutation would trigger undesirable re-renders in components that have no visual relevance to the current view.

### Changed
Refactored Tabs view and controller which led to a cascade of issues.  Multiple views and controllers rely on hooks, utils, etc. in the global/window namespace.

Refactored the stacking of hook pointers into global namespace (invariably leads to difficult-to-RCA issues).

- Start refactoring heavy components into reusable subcomponents.
- Start refactoring enums and UI constants into separate utils and constants lists.
- Start converting some components, utilities into type-safe typescript.

To help with ongoing reduction in tech-debt, code smells, added `react` and `react-hooks` eslint plugins that will callout issues wrt components and hooks usage.

### Fixed
Multiple chained renders into views not in scope for the desired activity (e.g., rendering CharacterPreview and sub components when working in the Relics tab).

## Checklist

- [x] I have added commit messages that are descriptive and meaningful.
- [x] I have tested the changes locally.
- [x] I have reviewed the code changes.

## Additional Notes
Largely non-breaking changes, but i am unaware of what a full regression test pass would involve.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,9 +5,6 @@
   "homepage": "https://fribbels.github.io/hsr-optimizer",
   "dependencies": {
     "@ant-design/icons": "^5.2.6",
-    "@testing-library/jest-dom": "^5.17.0",
```
Comment (cnojima) [inline package.json:8]: i'm hoping OOTB webpack treeshaking didn't ship testing harness code with the prod bundles...

File: src/App.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import React, { useEffect } from "react"
-import { ConfigProvider, Flex, Layout, message, notification, theme, Typography } from 'antd';
+import { ConfigProvider, Flex, Layout, message, /* notification,  */theme, Typography } from 'antd';
```
Comment (cnojima) [inline src/App.js:2]: `notification` seems like cruft - wasn't used anywhere in the app.  left it commented out, but it probably can be straight nuked.

File: src/App.js [hunk 3]
```diff
@@ -26,9 +26,22 @@ const App = () => {
     <ConfigProvider
```
Comment (cnojima) [inline src/App.js:26]: we had 4+ instances of `<ConfigProvider />` - had lots of UI/css spread out over several top & mid-level components. refactored to the top for consistency and maintainability.

File: src/components/CharacterTab.js [hunk 4]
```diff
@@ -184,7 +182,11 @@ export default function CharacterTab(props) {
     characterGrid.current.api.redrawRows()
     window.forceCharacterTabUpdate()
     Message.success('Successfully unequipped character')
-    global.relicsGrid.current.api.redrawRows()
+
+    // Char and Relic tab are never mounted together - this is unavailble
+    if (global.relicsGrid?.current?.api) {
```
Comment (cnojima) [inline src/components/CharacterTab.js:187]: this could probably be deleted?  with refactor, that api will never be mounted when view CharacterTab

**Comment by fribbels** [comment]
image

Still poking around but found this happens if I switch tabs too fast

File: src/App.js (hunk 2 above)
Comment (fribbels) [reply src/App.js:2]: Right, that's gone now

**Comment by fribbels** [comment]
Looks like a lot of things are getting reset - repro steps:

1. Double click rank 2 character
2. Press start then cancel, sort by speed
3. Switch to character tab and do nothing
4. Switch to optimizer tab
5. Selected char switched to rank 1 character, sort on grid got reset

**Comment by fribbels** [comment]
KelZ tab 'Import Relics + Characters' errors with 
image

**Comment by fribbels** [comment]
Scoring algorithm -> Save changes errors out with 
image

**Comment by cnojima** [comment]
Knocked out the scoring issue - i did modify the behavior in the Relics Tab to redraw on the new scoring algo.  with this change, you can remove the "Rescore" button.  deltas In my latest commit.

looking into state issues with char/relic grids.  i'm not familiar enough with how the app should/expects to work, so i'll have to compare to live site.

**Comment by cnojima** [comment]
Refactored focus/selected "character" logic in store.  Single point of reference for all views.  eliminated MUCH thrash/rerender.

**Comment by fribbels** [comment]
Couple issues im noticing: if I switch characters from the Characters tab it empties out the light cone options and clicking Start breaks the page

image

Clicking relic cards doesn't work on Optimizer & Character tab, though does work on Relics tab

**Comment by cnojima** [comment]
Addressed CharacterPreview -> Relic click -> Relic edit overlay in recent commit.

Still trying to reproduce LC empty > start issue.

**Comment by cnojima** [comment]
found and fixed the LC passives issue.  PR updated.

**Comment by fribbels** [comment]
I think this is a good enough state to merge in so it doesn't get blocked on other changes, there's a couple things that worry me like the multiple 100ms setTimeouts, but I'm cool with refactoring those later.

**Pull request merged**